idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionInfoProvider.kt

```diff
@@ -31,4 +39,29 @@ internal class KtFirExpressionInfoProvider(
         val firWhenExpression = whenExpression.getOrBuildFirSafe<FirWhenExpression>(analysisSession.firResolveState) ?: return emptyList()
         return FirWhenExhaustivenessTransformer.computeAllMissingCases(analysisSession.firResolveState.rootModuleSession, firWhenExpression)
     }
-}
\ No newline at end of file
+
+    override fun getLastExpressionIfImplicitReturn(ktLambdaExpression: KtLambdaExpression): KtExpression? = withValidityAssertion {
```

Comment:
Looks like, no FIR building is needed for your function, you can get the information you need by corresponding PSI.

```
ktLambdaExpression.functionLiteral.bodyExpression.statements.lastOrNull()
```

So, it can be implemented in your UAST code

Reply:
With the fix for retrieval of anonymous function symbol from `KtLambdaExpression` (or `KtFunctionLiteral`), yeah, we can see the last expression. But, in FE 1.0, binding context has a slice that indicates whether that last expression is used as a lambda return or not. For that case, FIR transforms the lambda body, in particular, that last expression with a wrapped implicit return expression. I still believe that is the most accurate information.

But, perhaps we can mimic what `FirAnonymousFunction#addReturn` in `FirDeclarationResolveTransformer` does in FIR UAST side: https://github.com/JetBrains/intellij-kotlin/pull/80/commits/34b3564f193b4bb533475d5abe0d0e55038e716e#diff-41b037ea4d3673022f43bfaeac2b249659baa8069275309ff27c7649ca07d394 It's not 100% correct yet, but I guess we can match the behavior eventually.

So, reverted all the additions to `KtExpressionInfoProvider`.